x/mint/keeper/emissions.go

```diff
@@ -17,39 +17,48 @@ func GetLockedVestingTokens(
 	blocksPerMonth uint64,
 	blockHeight math.Int,
 	params types.Params,
-) (total, preseedInvestors, investors, team math.Int) {
+	monthsUnlocked math.LegacyDec,
+) (total, preseedInvestors, investors, team math.Int, updatedMonthsUnlocked math.Int) {
 	// foundation is unlocked from genesis
 	// participants are unlocked from genesis
 	// investors and team tokens are locked on a 1 year cliff three year vesting schedule
 	blocksInAYear := math.NewIntFromUint64(blocksPerMonth * 12)
```

Comment:
Here, `blocksInAYear` would vary depending on the current value of `blocksPerMonth`. In a scenario where this goes down (or up) dramatically, it not only affects current month calculations but also the value for the whole year. 
Is this expected and covered?
E.g. it could mean that block `b` is `LT(blocksInAYear)` but block `b+1` is not.
Current setup is "estimating the past", isn't it?

Suggestion: we can keep track the block at which the monthUnlocks happened, so the yearly estimation does not extend "over the past", ie, we can keep track of what already happened, and only estimate what comes next.
A year can be thus computed as "last 11 months time + `blocksPerMonth`")

Reply:
`blocksInAYear` is only used in the if/else statement a couple lines below.

If block `b+1` is not less than blocksInAYear, but a previous block `b` went into the else branch because it was greater than blocksPerYear, the `updatedBlocksPerMonth` in the past would have been previously stored in the keeper later in `RecalculateTargetEmission` and so at the time of evaluation of block `b+1` in if branch `monthsUnlocked.LT(twelve)` would already be greater than 12.


We could track the specific block height but I think it makes more sense to just track the stored months already unlocked (monthsUnlocked)

Reply:
I think maybe the variable could be renamed to be better

Reply:
if the `blocksPerMonth` value decreases a lot, the `else` could be reached before `monthsAlreadyUnlocked` get to 12 (because it would not satisfy first condition). Is this ok? 
Wondering whether we should remove the yearly check `blockHeight.LT(blocksInAYear)` and just rely on monthly unlocks.

Another note:
since `monthsUnlocked` is based on a division over `blocksPerMonth`, which is variable, a decrease in this value could make `monthsUnlocked` to reach 36 faster which could make `monthsLocked := thirtySix.Sub(monthsAlreadyUnlocked)` to be 0 => in that case, will that mean all remaining tokens will be unlocked, or that no tokens will be unlocked at all?

Reply:
Hmmm reviewing this, could it be that this `blockHeight.LT(blocksInAYear)` condition is just a sort of "pre-estimation" of monthsAlreadyUnlocked?  ie `monthsAlreadyUnlocked` at that point marking the "past" (did we already reach 12 in the past?), while this condition marking the present (are we reaching it now under current yearly estimation numbers?)

In that case the mechanism looks ok - leaving the last note ab how a sudden change affects unlocks.

Reply:
yes, the code can be simplified, the if else statement is not necessary, since the math will work either way (the unlocked tokens will be zero before 12 months in the math). making changes to PR now

> since monthsUnlocked is based on a division over blocksPerMonth, which is variable, a decrease in this value could make monthsUnlocked to reach 36 faster 

Correct. In my opinion the right thing to do here is to always err on the side of saying the circulating supply is larger than it actually is in real life - this means that we err on the side of having too little inflation, rather than (potentially) too much inflation

This code doesn't actually "unlock" or send any tokens to these investors, it only is tracking how much they're supposed to have based on the unlock schedule. Our off-chain custodian partner responsible for the investor allocations will be the one legally responsible for introducing these tokens into the token supply by sending them to people who could use, transfer, or sell them